### Ομάδα 70: 
#### Χρήστος Παπαδημητρίου, el18017
#### Βαρθολομαίος Βαμβακάρης, el18072
#### Ζαχαρίας-Παύλος Αναστασιάδης, el18161

# Dataset : Connectionist Bench (Sonar, Mimes vs. Rocks)

In [6]:
!pip install --upgrade pip #upgrade pip package installer
!pip install scikit-learn --upgrade #upgrade scikit-learn package
!pip install numpy --upgrade #upgrade numpy package
!pip install --upgrade matplotlib # Κάνουμε update την matplotlib

Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable
Defaulting to user installation because normal site-packages is not writeable


## Εισαγωγή & Επισκόπηση

In [8]:
import pandas as pd

df = pd.read_csv('sonar.all-data', header=None)
df

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [21]:
# check if we have any missing values

isNullArray, nullValsPresent = df.isnull(), False

for numCol in isNullArray:
    if True in isNullArray[numCol]:
        print("Missing Values in column: " + str(col))
        flag = True

if not nullValsPresent:
    print("There are missing data in the Dataset")

There are missing data in the Dataset


#### Σύντομη Παρουσίαση του Dataset
Το παραπάνω Dataset περιλαμβάνει 208 δείγματα που αντιστοιχούν σε patterns συχνοτήτων. Αυτά τα patterns έχουν προκύψει από την αντανάκλαση σημάτων Sonar σε μεταλλικούς κυλίνδρους και σε πετρώματα. Συγκεκριμένα τα σήματα Sonar που χρησιμοποιήθηκαν είναι "Chirps" αυξανόμενης συχνότητας. Το Dataset περιλαμβάνει αποτελέσματα εκπομπής Sonar από ποικίλες γωνίες (aspect angles), μεχρί 90$^{o}$ για τους μεταλλικούς κυλίνδρους και μέχρι 180$^{o}$ για τα πετρώματα. 

Κάθε pattern περιγράφεται από 60 αριθμούς μεταξύ 0.0 και 1.0. Καθένας από αυτούς τους αριθμούς αντιστοιχεί στο ολοκλήρωμα της ενέργειας μιας συγκεκριμένης μπάντας συχνοτήτων για μια περίοδο στον χρόνο. 

Τα labels που συναντάμε στο dataset είναι 'R' για τα πετρώματα (Rocks) και 'Μ' για τους μεταλλικούς κυλίνδρους (Metal cylinder). 

Σημειώνεται ότι **δεν** χρειάστηκε να γίνουν μετατροπές στο αρχείο 'sonar.all-data' πριν την εισαγωγή του στο Notebook

#### Ακολουθούν κάποιες επιπλέον πληροφορίες για το Dataset:
* Το Dataset περιλαμβάνει 208 δείγματα. Κάθε δείγμα έχει 60 χαρακτηριστηκά. Το καθένα από αυτά είναι μία τιμή μεταξύ 0.0 και 1.1 (Το ολοκήρωμα της ενέργειας που αναφέραμε παραπάνω) 
* Δεν υπάρχουν επικεφαλίδες, ούτε αρίθμηση γραμμών
* Οι ετικέτες των κλάσεων παίρνουν τιμές 'R' και 'Μ', όπως αναφέραμε παραπάνω. Βρίσκονται στην 61$^{η}$ κολόνα (index 60).
* Δεν υπάρχουν δείγματα με απουσιάζουσες τιμές
* Η κλαση R έχει 97 δείγματα (ποσοστό: 46.6%), ενώ η κλάση M έχει 111 δείγματα (ποσοστό: 53.4%). Σύμφωνα με το όριο 60%-40% πρόκειται για ισορροπημένο Dataset.

## Προετοιμασία

In [40]:
# split the dataframe into features matrix and labels column
features_df, labels_df = df.iloc[:,1:60], df.iloc[:,[60]]
labels_df

60
0    R
1    R
2    R
3    R
4    R
..  ..
203  M
204  M
205  M
206  M
207  M

[208 rows x 1 columns]

In [41]:
# turn features into a numpy array
features_np = features_df.values
features_np

array([[0.0371, 0.0428, 0.0207, ..., 0.0084, 0.009 , 0.0032],
       [0.0523, 0.0843, 0.0689, ..., 0.0049, 0.0052, 0.0044],
       [0.0582, 0.1099, 0.1083, ..., 0.0164, 0.0095, 0.0078],
       ...,
       [0.0437, 0.018 , 0.0292, ..., 0.0138, 0.0077, 0.0031],
       [0.0353, 0.049 , 0.0608, ..., 0.0079, 0.0036, 0.0048],
       [0.0363, 0.0136, 0.0272, ..., 0.0036, 0.0061, 0.0115]])

In [48]:
# replace (R,M) with (0,1) in labels 

labels_df = labels_df.replace({'R':0}).replace({'M':1})
labels_np = labels_df.values.flatten()
labels_np

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [50]:
from sklearn.model_selection import train_test_split

# split our data into training and testing data
train, test, train_labels, test_labels = train_test_split(features_np, labels_np, test_size=0.30)

array([[0.1163, 0.0866, 0.0358, ..., 0.0141, 0.0068, 0.0086],
       [0.0644, 0.039 , 0.0173, ..., 0.016 , 0.0095, 0.0011],
       [0.0078, 0.0092, 0.0387, ..., 0.0049, 0.0024, 0.0016],
       ...,
       [0.0436, 0.0447, 0.0844, ..., 0.0077, 0.0246, 0.0198],
       [0.044 , 0.0137, 0.0084, ..., 0.0081, 0.0086, 0.0095],
       [0.121 , 0.1268, 0.1498, ..., 0.0101, 0.0228, 0.0124]])

## Ταξινόμηση

## Αποτελέσματα & Συμπεράσματα